src/sage/rings/species.py

```diff
@@ -0,0 +1,475 @@
+    return GAP_FAIL != libgap.RepresentativeAction(G, H1, H2)
+
+
+class ElementCache():
+    def __init__(self):
+        r"""
+        Class for caching elements.
+        """
+        self._cache = dict()
+
+    def _cache_get(self, elm):
+        r"""
+        Return the cached element, or create it
+        if it doesn't exist.
+
+        ``elm`` must implement the following methods:
+          ``_element_key`` - hashable type for dict lookup.
+          ``__eq__`` - to compare two elements.
+        """
+        key = elm._element_key()
+        if key in self._cache:
+            lookup = self._cache[key]
+            for other_elm in lookup:
+                if elm == other_elm:
+                    return other_elm
+            else:
+                lookup.append(elm)
+        else:
+            self._cache[key] = [elm]
+        return elm
+
+
+class ConjugacyClassOfDirectlyIndecomposableSubgroups(Element):
+    def __init__(self, parent, C, ambient_group):
+        r"""
+        A conjugacy class of directly indecomposable subgroups.
+        """
+        Element.__init__(self, parent)
+        self._C = C
+        self._smallgen = C.gens_small()
+        self._ambient_group = ambient_group
+
+    @cached_method
+    def subgroup_of(self):
+        r"""
+        Return the group which this conjugacy class of
+        directly indecomposable subgroups belongs to.
+        """
+        return self._ambient_group
+
+    @cached_method
+    def _element_key(self):
+        r"""
+        Return the key for this element.
+        """
+        return tuple([self._C.degree(), self._C.order(),
+                      self._ambient_group])
+
+    def __hash__(self):
+        return hash(self._element_key)
+
+    def _repr_(self):
+        r"""
+        Return a string representation of ``self``.
+        """
+        return f"{self._smallgen}"
+
+    def __eq__(self, other):
+        r"""
+        Return if ``self`` is equal to ``other``.
+
+        Two directly indecomposable subgroups belong to the same conjugacy
+        class if they have the same ambient group and are conjugate within it.
+        """
+        # No __le__ method because does it really make sense to have one here?
+        return (isinstance(other, ConjugacyClassOfDirectlyIndecomposableSubgroups)
+                and self._ambient_group == other._ambient_group
+                and _is_conjugate(self._ambient_group, self._C, other._C))
+
+class ConjugacyClassesOfDirectlyIndecomposableSubgroups(UniqueRepresentation, Parent, ElementCache):
+
+    def __init__(self):
+        r"""
+        Conjugacy classes of directly indecomposable subgroups.
+        """
+        Parent.__init__(self, category=InfiniteEnumeratedSets())
+        ElementCache.__init__(self)
+    
+    def _element_constructor_(self, x):
+        r"""
+        ``x`` is an element of ``self`` or a tuple `(H, G)` such that
+        `H` is directly indecomposable and `G` is the ambient group of
+        `H`.
```

Comment:
If you really want to handle different ambient groups, the ambient group should be a parameter of `__init__`.

It is correct that we can have two directly indecomposable groups $F$ and $G$, which are conjugate with respect to $\mathfrak S_n$, but not conjugate in some other ambient group $H < \mathfrak S_n$.  However, we really only need to consider conjugacy with respect the symmetric group.